Add doc section about performance considerations (#2115)
cocotb/cocotb: documentation/source/best_practices.rst

.. _best_practices:

**************
Best Practices
**************

.. _performance:

Performance Considerations
==========================

In order to get the maximum performance out of a cocotb testbench,
the following points may help you.

Clock Generator in HDL
----------------------

Generate your clock with a HDL module instead of using :class:`~cocotb.clock.Clock`.

> Doing so prevents calls back into Python every single clock half-period.

Reviewer: It might be worth expanding this. For example, what are best practices about how to implement that? I'd recommend they use a wrapper instead of adding behavior clock generation to their design. And that has implications, `TOPLEVEL` will have to change, and so the dut parameter will change as well.

Maybe that's all obvious?

Author: Not obvious, we should actually have a whole example for this.
I'll see what I can do.

Performance Impact: **Major**

Reusing Objects
---------------

Do not create single-use objects in a loop but instead create them outside the loop and reuse them.

> Typical cases are :class:`~cocotb.triggers.Timer` or :class:`~cocotb.clock.Clock` objects as shown below:

Reviewer: ```suggestion
Typical cases are :class:`~cocotb.triggers.Timer` or :class:`~cocotb.triggers.RisingEdge` of the clock signal, as shown below:
```

You shouldn't create multiple `cocotb.clock.Clock`s at all, but `RisingEdge(clk)` is very common.

Author: Indeed, changed.

.. list-table::

   * - .. code-block:: python
          :caption: Inefficient ``Timer`` object re-creation

          while True:
              await Timer(0.5, "ns")

>               signal <= ~signal

Reviewer: ```suggestion
              signal <= ~signal.value
```

the handle class does not have the `__invert__` overload

Author: Ah, that's a small bug also in our existing docs then.

Author: Done, also in ``coroutines.rst``.

     - .. code-block:: python
          :caption: More efficient ``Timer`` object re-use

          halfperiod = Timer(0.5, "ns")
          while True:
              await halfperiod

>               signal <= ~signal

Reviewer: ```suggestion
              signal <= ~signal.value
```
